src/libYARP_OS/src/Timer.cpp

```diff
@@ -0,0 +1,48 @@
+#include <yarp/os/Timer.h>
+#include <yarp/os/Time.h>
+using namespace yarp::os;
+
+Timer::Timer(const double rate, TimerCallback callback, Mutex* mutex, const bool inputOneshot) :
+    RateThread(rate),
+    m_callback(callback),
+    m_oneshot(inputOneshot),
+    m_mutex(mutex)
+{}
+
+
+bool Timer::threadInit()
+{
+    m_startStamp = yarp::os::Time::now();
+    return true;
+}
+
+void Timer::run()
+{
+    if(m_runCount == 0)
+    {
+        m_runCount++;
+        return;
+    }
+
+    YarpTimerEvent event;
+
+    event.currentReal     = yarp::os::Time::now();
+    event.currentExpected = m_startStamp + m_runCount * getRate() * 0.001;
+    event.lastExpected    = event.currentExpected - getRate() * 0.001;
+    event.lastReal        = m_lastReal;
+    event.lastDuration    = event.currentReal - m_lastReal;
+    event.runCount        = m_runCount;
+
+    if(m_mutex) m_mutex->lock();
+
+    m_callback(event);
+
+    if(m_mutex) m_mutex->unlock();
+
+    m_lastReal = event.currentReal;
+    if(m_oneshot)
+    {
+        askToStop();
+    }
+    m_runCount ++;
```

Comment:
Nitpick on style... you used `m_runCount++` and `m_runCount ++` here. I'd stick with the first.